### PR TITLE
feat(cli): harden CLI input validation

### PR DESCRIPTION
**Pull request by sjinks**
## Summary

This PR hardens CLI replacement-argument validation to reject malformed or unsafe inputs before any stream processing starts, while preserving valid migration patterns for WordPress SQL search-replace use cases.

It also tightens expansion-safety checks to prevent pathological growth across ordered replacement chains, and follows up on review feedback to improve usage/error semantics and boundary-match correctness.

Issue reference: [PLTFRM-2239](https://linear.app/a8c/issue/PLTFRM-2239/insufficient-input-validation)

## Changes

- Added structured replacement-argument validation with explicit error categories and exit-code mapping for usage errors vs invalid `<from>` vs invalid `<to>`.
- Added strict input-form validation for accepted argument shapes (bare token, scheme token, host with optional port/path, or full http/https URL) and rejected unsupported patterns (e.g. traversal, query/fragment/userinfo, malformed host/port forms, serialization markers).
- Added argument-count and size guardrails, duplicate/no-op replacement rejection, and expansion controls.
- Added/updated ordered-chain expansion checks, including seam-assembly handling and a per-split seam predicate fix from review feedback.
- Updated CLI usage text to reflect repeated pair syntax.
- Updated README language for expansion policy (including 16x cap and ordered-chain behavior).
- Added extensive test coverage for validation behavior, exit-code semantics, expansion-chain safety, seam-assembly edge cases, and read-error handling path compatibility from trunk.

## Testing and Validation

- `go test -count=1 ./...` (pass)
- `go vet ./...` (clean)

## Review Notes

- Copilot review comments were addressed in follow-up commits.
- Boundary-overlap logic was revised twice based on review: first to remove byte-set overcounting, then to enforce a single-split seam-match condition.
- All review threads are currently marked resolved.

## Risks and Follow-up

- Validation is intentionally stricter and may reject inputs that were previously tolerated but were ambiguous or unsafe.
- Exit-code handling is now more semantically aligned; downstream scripts that depended on prior misclassified failures should verify expectations.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -120,19 +154,363 @@ func main() {
...
+	return validateInput(in, length) == nil
+}
+
+func validateReplacementArgs(args []string) error {
+	if len(args)%2 > 0 {
+		return newReplacementArgError(exitUsage, "All replacements must have a <from> and <to> value")
+	}
+
+	if len(args)/2 > maxReplacementPairs {
+		return newReplacementArgError(exitInvalidFrom, "Too many replacement pairs: maximum is %d", maxReplacementPairs)
```
[the remark this answers is not in the document]
Comment (sjinks) [reply main.go]: Good catch — addressed in 47d0015. The "Too many replacement pairs" branch in `validateReplacementArgs` now returns `exitUsage` instead of `exitInvalidFrom`, and `TestValidationExitCodes` has a new "too many replacement pairs" case that asserts `wantCode: exitUsage` and the expected stderr message. `TestTooManyReplacementPairs` is unchanged. `go vet ./...`, `go build ./...`, and `go test -count=1 ./...` all pass locally.

File: main.go [hunk 2]
```diff
@@ -5,28 +5,74 @@ import (
...
 
 const (
-	badInputRe   = `\w:\d+:`
-	inputRe      = `^[A-Za-z0-9_\-\.:/]+$`
-	minInLength  = 4
-	minOutLength = 2
+	badInputRe          = `\w:\d+:`
+	inputRe             = `^[A-Za-z0-9_\-\.:/]+$`
+	safePartRe          = `^[A-Za-z0-9_.-]+$`
+	minInLength         = 4
```
[the remark this answers is not in the document]
Comment (sjinks) [reply main.go:24]: Thanks, but I don't think this one is accurate. The regex on line 24 is:

```go
safePartRe = `^[A-Za-z0-9_.-]+$`
```

The character class `[A-Za-z0-9_.-]` explicitly includes `_` (underscore), `.`, and `-`, in addition to alphanumerics — so `my_plugin` and similar underscore-bearing tokens/path segments are already accepted by both `validateBareInput` and `validatePath`. This is also exercised by existing tests (e.g. inputs like `my_plugin_name` / `another_plugin` in `main_test.go`), which would fail if `_` were rejected.

Looks like the suggestion is based on a misread of the character class. No change needed here.

File: main.go [hunk 3]
```diff
@@ -120,19 +154,363 @@ func main() {
...
+		bytesInLaterFrom[laterFrom[index]] = true
+	}
+
+	count := 0
+	for index := 0; index < len(current); index++ {
+		if bytesInLaterFrom[current[index]] {
+			count++
+		}
+	}
+
```
[the remark this answers is not in the document]
Comment (sjinks) [reply main.go]: Good catch — addressed in 8644f58. You were right that the helper didn't compute boundary overlap; it counted every byte in `current` whose value appeared in `laterFrom`'s byte set, which is unrelated to whether `laterFrom` can be assembled across the seam between adjacent copies of `current`. Closer inspection also showed that the downstream `matchCount`/`originalLength` math the helper fed never produced a reachable rejection under the existing guards, so the whole boundary check was effectively inert — masking a real attack regime (non-expanding `earlier` + expanding `later` whose `from` matches across the seam).

The new design:

- Removed `aggregateBoundaryOverlapCount`.
- Added `boundarySeamAssemblesMatch(current, laterFrom)`, which reports whether there exists `i ∈ [1, L-1]` such that `current` ends with `laterFrom[:i]` and `current` begins with `laterFrom[i:]`. This is exactly `suffixOverlap + prefixOverlap >= L`, where each overlap is the largest `k` such that `current` ends with `laterFrom[:k]` (resp. begins with `laterFrom[L-k:]`).
- Rewrote `validateBoundaryExpansionChain` to gate on that predicate and then compare the amortized output per input copy of `earlier.from`, i.e. `len(current) + len(later.to) - len(later.from)`, against `len(earlier.from) * maxExpansionFactor`.

Regression coverage added to `TestValidateReplacementArgs`:

- `boundary helper accepts shared bytes without affix overlap` — locks in the fix for the false-positive overcounting (`current="bbbbbbbbbb"`, `later.from="cccbbbbbcccbbbbccc"`: alphabets intersect on `'b'` but neither suffix nor prefix overlap is nonzero, so the new helper correctly accepts).
- `boundary expansion rejected when seam assembly compounds` — genuine seam assembly past the 16× amortized budget (`current="abab"`, `later.from="abababab"`, `later.to` length 128) is rejected.
- `boundary expansion accepted when seam assembly within budget` — exact-threshold tightness (`current="abababab"`, `later.to` length 128, ratio = 16) is accepted.

`gofmt`, `go vet ./...`, `go build ./...`, `go test -count=1 ./...` all clean.

File: main.go [hunk 4]
```diff
@@ -37,36 +83,24 @@ func main() {
 		return
 	}
 
-	if len(os.Args) < 3 {
+	args := flag.Args()
+
+	if len(args) < 2 {
 		fmt.Fprintln(os.Stderr, "Usage: search-replace <from> <to>")
```
[the remark this answers is not in the document]
Comment (sjinks) [reply main.go]: Addressed in ac21b26. The usage line is now:

```go
fmt.Fprintln(os.Stderr, "Usage: search-replace <from> <to> [<from> <to> ...]")
```

so it reflects that the CLI accepts up to `maxReplacementPairs` (currently 64) pairs per invocation, matching what `validateReplacementArgs` enforces.

File: README.md [hunk 5]
```diff
@@ -24,6 +24,15 @@ be used in domain names. Since the most common usage for search-replace is
...
 
+CLI replacement values may be bare tokens such as `sections` or `https`, scheme
+tokens `http:` and `https:`, bare hosts/domains with optional numeric ports and
+safe path segments such as `example.com:8080/wp-content`, or full `http://` and
+`https://` URLs with a host. Filesystem-style paths, traversal segments,
+unsupported schemes, userinfo, query strings, and fragments are rejected before
+input processing begins.
+
+At most one replacement pair may expand its input per invocation.
+
```
[the remark this answers is not in the document]
Comment (sjinks) [reply README.md:40]: Addressed in 8374d3a. The "Considerations" section now documents all three facts about the expansion policy:

- the 16× per-pair cap (each `<to>` may be at most 16× the byte length of its paired `<from>`),
- the single-expanding-pair rule per invocation,
- and that the same 16× factor bounds cumulative expansion across the ordered replacement chain, including boundary effects between adjacent replacements.

File: main.go [hunk 6]
```diff
@@ -120,19 +154,389 @@ func main() {
...
+
+	prefixOverlap := 0
+	for k := maxK; k > 0; k-- {
+		if strings.HasPrefix(current, laterFrom[L-k:]) {
+			prefixOverlap = k
+			break
+		}
+	}
+
+	return suffixOverlap+prefixOverlap >= L
```
[the remark this answers is not in the document]
Comment (sjinks) [reply main.go]: Good catch — addressed in d40f2bf. You're right that `suffixOverlap + prefixOverlap >= L` over independently-maximized overlaps doesn't imply a single-split seam match. Your `current="cdefabcd"`, `laterFrom="abcdef"` example is the cleanest demonstration: max suffix is `"abcd"` at split `i=4`, max prefix is `"cdef"` at split `i=2`, the sum is 8 ≥ L=6 so the old predicate returned `true`, yet no single `i ∈ [1,5]` simultaneously satisfies both side conditions, so `"abcdef"` cannot actually be assembled across the seam.

The fix:

- `main.go`: `boundarySeamAssemblesMatch` now iterates `i ∈ [1, L-1]` and returns true only when a single `i` satisfies both `strings.HasSuffix(current, laterFrom[:i])` and `strings.HasPrefix(current, laterFrom[i:])`. The doc comment was updated to describe the per-split semantics. `validateBoundaryExpansionChain` and surrounding code are unchanged. Naive O(L · len(current)) is comfortable under the existing argument-length and pair-count caps.
- `main_test.go`: added `TestBoundarySeamAssemblesMatch`, a direct table-driven unit test on the helper. Your counterexample is included as a `want: false` case (it would fail against the old sum-of-maxima implementation, locking in the regression). Other cases cover genuine seam assembly, full equality, disjoint alphabets, shared bytes without affix overlap, single-byte `laterFrom`, and empty `current`.

`gofmt`, `go vet ./...`, `go build ./...`, and `go test -count=1 ./...` all clean.
